proposals/0186-loaded-transaction-data-size-specification.md

```diff
@@ -0,0 +1,163 @@
+
+* BPF loaders required by instructions' program IDs are counted against
+transaction data size. BPF loaders required by CPI programs are not. If a
+required BPF loader is also included in the accounts list, it is counted twice.
+* The size of a program owned by LoaderV3 may or may not include the size of its
+programdata depending on how the program account is used on the transaction.
+Programdata is also itself counted if included in the transaction accounts list.
+This means programdata may be counted zero, one, or two times per transaction.
+* Due to certain quirks of implementation, loader-owned accounts which do not
+contain valid programs for execution may or may not be counted against the
+transaction data size total depending on how they are used on the transaction.
+This includes, but is not limited to, LoaderV3 buffer accounts, and accounts
+which fail ELF validation.
+* Accounts can be included on a transaction account list without being an
+instruction account, fee-payer, or program ID. These accounts are presently
+loaded and counted against transaction data size, although they can never be
+used for any purpose by the transaction.
+
+All validator clients must arrive at precisely the same transaction data size
+for all transactions because a difference of one byte can determine whether a
+transaction is executed or failed, and thus affects consensus. Also, we want the
+calculated transaction data size to correspond well with the actual amount of
+data the transaction requests.
+
+Therefore, this SIMD seeks to specify an algorithm that is straightforward to
+implement in a client-agnostic way, while also accurately accounting for all
+account data required by the transaction.
+
+## New Terminology
+
+No new terms are introduced by this SIMD, however we define these for clarity:
+
+* Instruction account: an account passed to an instruction in its accounts
+array, which allows the program to view the actual bytes contained in the
+account. CPI can only happen through programs provided as instruction accounts.
+* Transaction accounts list: all accounts for the transaction, which includes
+instruction accounts, the fee-payer, program IDs, and any extra accounts added
+to the list but not used for any purpose.
+* LoaderV3 program account: an account owned by
+`BPFLoaderUpgradeab1e11111111111111111111111` which contains in its account data
+the first four bytes `02 00 00 00` followed by a pubkey which points to an
+account which is defined as the program's programdata account.
+
+For the purposes of this SIMD, we make no assumptions about the contents of the
+programdata account.
+
+## Detailed Design
+
+The proposed algorithm is as follows:
+
+1. Given a transaction, take the unique set of account keys which are used as:
+    * An instruction account.
+    * A program ID for an instruction.
+    * The fee-payer.
+2. Each account's size is determined solely by the byte length of its data prior
+to transaction execution.
```

Comment:
iirc, original motivation is primarily focus on memory footprint and how that chunk of memory is used, loading from accountdb and serializing wasn't discussed much then, it might make sense now

Reply:
How about 96 bytes? From https://github.com/anza-xyz/agave/blob/4e7f7f76f453e126b171c800bbaca2cb28637535/programs/bpf_loader/src/serialization.rs#L417, I see we have to serialize 3 bytes for `is_writable`, `is_signer`, and `is_executable`, 4 bytes for original data length, 8 bytes each for `data length`, `lamports`, and `rent_epoch`, 32 bytes for the `account_key`, and 32 bytes for the `owner` and we adjust for alignment to the nearest multiple of 8. cc @Lichtso

Reply:
Yes that is the size of the serialized account metadata. But have you thought about the resize / realloc padding (`MAX_PERMITTED_DATA_INCREASE`)?

Reply:
Will that be used in loaderv4 too?

Reply:
Yes, serialization will stay ABIv1 in loader-v4.
ABIv2 is its own [SIMD](https://github.com/solana-foundation/solana-improvement-documents/pull/177).

Reply:
realloc headroom is quite different conceptually from "loaded data" tho isnt it? space is reserved, but no data is loaded

Reply:
Sounds like we need a (loose?) definition for what loaded data is intended to mean. It could be the amount of bytes read from accounts db (in that case maybe programdata is only counted once?) or it could be roughly the amount of bytes loaded into memory before tx execution?

In either case I think it makes sense to include the account metadata overhead of approx. 64 bytes and exclude the overhead of the account key itself and the realloc buffer since they're not really loaded in any sense.

Reply:
im equally fine counting programdata once or twice, and after this pr lets us refactor we should only have to actually load it once. main concerns are that the algorithm is very simple implement correctly, and always counts programdata if you use the program

maybe a new definition (counting programdata once) like:
1. Given a transaction, take the unique set of account keys which are used as:
    * An instruction account.
    * A program ID for an instruction.
    * The fee-payer.
2. For all LoaderV3 program accounts, add the key of the programdata account it references to the set of account keys, if it does not already exist.
3. Each account's size is defined as the byte length of its data prior to transaction execution, plus 64 bytes to account for metadata. This is irrespective of how the account is used on the transaction.
4. The total transaction loaded account data size is the sum of these sizes.

Reply:
I'm also fine either way, slight bias to counting it once now if we only need to load it once from accounts-db. Do we agree on 64 bytes overhead?